controllers/metal3.io/hostfirmwarecomponents_test.go

```diff
@@ -0,0 +1,331 @@
+					Updates: []metal3api.FirmwareUpdate{
+						{
+							Component: "bmc",
+							URL:       "https://myurls/newbmcfirmware",
+						},
+					},
+					Components: []metal3api.FirmwareComponentStatus{
+						{
+							Component:          "bmc",
+							InitialVersion:     "1.0.0",
+							CurrentVersion:     "1.1.0",
+							LastVersionFlashed: "1.1.0",
+						},
+						{
+							Component:      "bios",
+							InitialVersion: "1.0.1",
+							CurrentVersion: "1.0.1",
+						},
+					},
+				},
+			},
+			ExpectedComponents: &metal3api.HostFirmwareComponents{
+				Spec: metal3api.HostFirmwareComponentsSpec{
+					Updates: []metal3api.FirmwareUpdate{
+						{
+							Component: "bmc",
+							URL:       "https://myurl/newbmcfirmware",
+						},
+					},
+				},
+				Status: metal3api.HostFirmwareComponentsStatus{
+					Updates: []metal3api.FirmwareUpdate{
+						{
+							Component: "bmc",
+							URL:       "https://myurl/newbmcfirmware",
+						},
+					},
+					Components: []metal3api.FirmwareComponentStatus{
+						{
+							Component:          "bmc",
+							InitialVersion:     "1.0.0",
+							CurrentVersion:     "1.1.0",
+							LastVersionFlashed: "1.1.0",
+						},
+						{
+							Component:      "bios",
+							InitialVersion: "1.0.1",
+							CurrentVersion: "1.0.1",
+						},
+					},
+					Conditions: []metav1.Condition{
+						{Type: "ChangeDetected", Status: "True", Reason: "OK"},
+						{Type: "Valid", Status: "True", Reason: "OK"},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.Scenario, func(t *testing.T) {
+			ctx := context.TODO()
+			prov := getMockHFCProvisioner(getCurrentComponents())
+
+			tc.ExpectedComponents.TypeMeta = metav1.TypeMeta{
+				Kind:       "HostFirmwareComponents",
+				APIVersion: "metal3.io/v1alpha1"}
+			tc.ExpectedComponents.ObjectMeta = metav1.ObjectMeta{
+				Name:            "hostName",
+				Namespace:       "hostNamespace",
+				ResourceVersion: "2"}
+
+			hfc := tc.CurrentHFCResource
+			r := getTestHFCReconciler(hfc)
+			// Create a bmh resource needed by hfc reconciler
+			bmh := createBaremetalHostHFC()
+
+			info := &rhfcInfo{
+				ctx: ctx,
+				log: logf.Log.WithName("controllers").WithName("HostFirmwareComponents"),
+				hfc: tc.CurrentHFCResource,
+				bmh: bmh,
+			}
+
+			err := r.updateHostFirmware(info)
+			assert.Equal(t, nil, err)
+
+			components, err := prov.GetFirmwareComponents()
+			assert.Equal(t, nil, err)
+			err = r.updateHostFirmwareComponents(components, info)
+			assert.Equal(t, nil, err)
+
+			// Check that resources get created or updated
+			key := client.ObjectKey{
+				Namespace: hfc.ObjectMeta.Namespace, Name: hfc.ObjectMeta.Name}
+			actual := &metal3api.HostFirmwareComponents{}
+			err = r.Client.Get(ctx, key, actual)
+			assert.Equal(t, nil, err)
+
+			// Trying to figure out how to check if the Status.Components also got updated.
+			assert.Equal(t, tc.ExpectedComponents.Status.Components, actual.Status.Components)
```

Comment:
Missing check on conditions. Did you intend to compare the whole Status?

Reply:
I plan to compare the whole Status, my idea is to check each one (Updates, Components, LastUpdated and Conditions)